src/graph.py

```diff
@@ -0,0 +1,156 @@
+"""Implement a class for a graph data structure."""
+
+
+class Graph(object):
+    """Graph class."""
+
+    def __init__(self):
+        """Initialize a graph."""
+        self._graph = {}
+
+    def nodes(self):
+        """Return a list of all nodes in graph."""
+        return list(self._graph.keys())
+
+    def edges(self):
+        """Return a list of edges in graph."""
+        edges = {}
+        for key in self._graph:
+            for i in self._graph[key]:
+                edges[(key, i)] = self._graph[key][i]
+        return edges
+
+    def add_node(self, val):
+        """Add a node with value of val to graph."""
+        self._graph[val] = {}
+
+    def add_edge(self, val1, val2, weight=0):
+        """Add a new edge to graph between val1 & val2 as well as add vals."""
+        if val1 in self._graph and val2 in self._graph:
+            if val2 not in self._graph[val1]:
+                self._graph[val1][val2] = weight
+        elif val1 in self._graph and val2 not in self._graph:
+            self._graph[val2] = {}
+            self._graph[val1][val2] = weight
+        elif val2 in self._graph and val1 not in self._graph:
+            self._graph[val1] = {}
+            self._graph[val1][val2] = weight
+        else:
+            self._graph[val1] = {val2: weight}
+            self._graph[val2] = {}
+
+    def del_node(self, val):
+        """Delete node w/val from graph, raises exception if not exist."""
+        if val in self._graph:
+            del self._graph[val]
+            for key in self._graph:
+                for i in self._graph[key]:
+                    if i == val:
+                        del self._graph[key]
```

Comment:
WHOA NOW! You're going to delete the whole node if the value you're looking for is found amongst its neighbors?!